File: frontend/src/types/graph.ts
/**
 * TypeScript types for Neo4j graph entities.
 *
 * These types mirror the Pydantic schemas from backend/src/schemas/graph.py.
 * All regulatory entities include citation fields for source traceability.
 */

/**
 * Confidence score for regulatory claims.
 */
export enum ConfidenceLevel {
  HIGH = 'high',
  MEDIUM = 'medium',
  LOW = 'low',
}

/**
 * Categories of government processes.
 */
export enum ProcessCategory {
  PERMITS = 'permits',
  LICENSES = 'licenses',
  BENEFITS = 'benefits',
}

/**
 * Types of web resources.
 */
export enum WebResourceType {
  HOW_TO = 'how_to',
  PROGRAM = 'program',
  PORTAL = 'portal',
  PDF_FORM = 'pdf_form',
}

/**
 * Status of a user application.
 */
export enum ApplicationStatus {
  PENDING = 'pending',
  APPROVED = 'approved',
  DENIED = 'denied',
}

/**
 * Category of application.
 */
export enum ApplicationCategory {
  NEW = 'new',
  RENEWAL = 'renewal',
  REPLACEMENT = 'replacement',
  RENTAL = 'rental',
}

/**
 * Base interface for timestamp fields.
 */
export interface TimestampFields {
  /** Timestamp when record was created (ISO 8601) */
  created_at: string
  /** Timestamp when record was last updated (ISO 8601) */
  updated_at: string
}

/**
 * Base interface for citation/provenance fields.
 * All regulatory data must be traceable to official sources.
 */
export interface CitationFields {
  /** URL to official source (regulation, webpage, PDF) */
  source_url: string
  /** Date when information was last verified (YYYY-MM-DD) */
  last_verified: string
  /** Confidence level in this data */
  confidence: ConfidenceLevel
}

/**
 * High-level government service or process.
 * Example: "Boston Resident Parking Permit"
 */
export interface Process extends CitationFields, TimestampFields {
  /** Unique process identifier (e.g., "boston_rpp") */
  process_id: string
  /** Human-readable process name */
  name: string
  /** One-sentence summary of the process */
  description: string
  /** Process category */
  category: ProcessCategory
  /** Governing authority (e.g., "City of Boston") */
  jurisdiction: string
}

/**
 * Actionable task within a process.
 * Example: "Gather proof of residency"
 */
export interface Step extends CitationFields, TimestampFields {
  /** Unique step identifier (e.g., "rpp.gather_proof") */
  step_id: string
  /** Parent process ID */
  process_id: string
  /** Short step label */
  name: string
  /** Full action description */
  description: string
  /** Sequence number (1-indexed) */
  order: number
  /** Official time estimate in minutes */
  estimated_time_minutes: number | null
  /** Median time from user feedback */
  observed_time_minutes: number | null
  /** Cost in USD */
  cost_usd: number
  /** Whether this step can be skipped */
  optional: boolean
}

/**
 * Eligibility condition or rule.
 * Example: "MA registration at Boston address"
 */
export interface Requirement extends CitationFields, TimestampFields {
  /** Unique requirement identifier (e.g., "rpp.req.ma_registration") */
  requirement_id: string
  /** Human-readable requirement description */
  text: string
  /** Reference to Facts Registry (e.g., "rpp.eligibility.registration_state") */
  fact_id: string
  /** Process ID this requirement applies to */
  applies_to_process: string
  /** Whether this blocks progress if not met */
  hard_gate: boolean
  /** Page or PDF section reference */
  source_section: string | null
}

/**
 * Atomic regulatory fact.
 * Example: "Proof of residency ≤30 days"
 */
export interface Rule extends CitationFields, TimestampFields {
  /** Unique rule identifier (e.g., "RPP-15-4A") */
  rule_id: string
  /** Exact regulation text */
  text: string
  /** Reference to Facts Registry */
  fact_id: string
  /** Scope of rule (e.g., "general", "rental", "military") */
  scope: string
  /** Section/page number reference */
  source_section: string
  /** When this rule took effect (YYYY-MM-DD) */
  effective_date: string | null
}

/**
 * Template for accepted documents.
 * Example: "Utility bill ≤30 days"
 */
export interface DocumentType extends CitationFields, TimestampFields {
  /** Unique document type identifier (e.g., "proof.utility_bill") */
  doc_type_id: string
  /** Human-readable document type name */
  name: string
  /** Maximum age in days (e.g., 30) */
  freshness_days: number
  /** Whether document name must match applicant name */
  name_match_required: boolean
  /** Whether document address must match Boston address */
  address_match_required: boolean
  /** List of example issuers (e.g., ["National Grid", "Eversource"]) */
  examples: string[]
}

/**
 * User-provided document instance.
 * Note: No citation fields - this is user data, not regulatory data.
 * Documents are auto-deleted after 24 hours.
 */
export interface Document extends TimestampFields {
  /** Unique document ID (UUID) */
  doc_id: string
  /** Reference to DocumentType */
  doc_type_id: string
  /** Document issuer (from OCR) */
  issuer: string
  /** Issue date (from OCR, YYYY-MM-DD) */
  issue_date: string
  /** Name on document (from OCR) */
  name_on_doc: string
  /** Address on document (from OCR) */
  address_on_doc: string
  /** File path (S3/local, deleted after 24h) */
  file_ref: string
  /** Whether document passed validation */
  verified: boolean
  /** Validation errors if any */
  validation_errors: string[]
  /** When document will be/was deleted (created_at + 24h, ISO 8601) */
  deleted_at: string | null
}

/**
 * Physical location for in-person steps.
 * Example: "Office of the Parking Clerk"
 */
export interface Office extends CitationFields, TimestampFields {
  /** Unique office identifier (e.g., "parking_clerk") */
  office_id: string
  /** Office name */
  name: string
  /** Full street address */
  address: string
  /** Room number */
  room: string | null
  /** Operating hours (e.g., "Mon-Fri, 9:00-4:30") */
  hours: string
  /** Phone number */
  phone: string | null
  /** Email address */
  email: string | null
}

/**
 * Boston parking neighborhood (RPP-specific).
 */
export interface RPPNeighborhood extends CitationFields, TimestampFields {
  /** Unique neighborhood identifier (e.g., "back_bay") */
  nbrhd_id: string
  /** Neighborhood display name */
  name: string
  /** Next auto-renewal audit date (YYYY-MM-DD) */
  auto_renew_cycle: string | null
  /** Street names with RPP signage */
  posted_streets: string[]
  /** Special rules or notes */
  notes: string | null
}

/**
 * Official web resource (page, PDF, portal).
 * Note: No citation fields - this IS a source, not derived from one.
 */
export interface WebResource extends TimestampFields {
  /** Unique resource identifier (e.g., "howto") */
  res_id: string
  /** Page title */
  title: string
  /** Full URL (must be unique) */
  url: string
  /** Resource type */
  type: WebResourceType
  /** Owning office/department (e.g., "Parking Clerk", "BTD") */
  owner: string
  /** Last successful fetch date (YYYY-MM-DD) */
  last_seen: string
  /** SHA256 hash of content for change detection */
  hash: string
}

/**
 * User (optional, for tracking).
 * Note: Minimal PII - only for Phase 2+.
 */
export interface Person {
  /** Unique person ID (UUID) */
  person_id: string
  /** Hashed email address */
  email: string
  /** When user record was created (ISO 8601) */
  created_at: string
}

/**
 * User's process session (Phase 2+).
 */
export interface Application extends TimestampFields {
  /** Unique application ID (UUID) */
  app_id: string
  /** Process ID (e.g., "boston_rpp") */
  process_id: string
  /** Application category */
  category: ApplicationCategory
  /** When application was submitted (ISO 8601) */
  submitted_on: string | null
  /** Current application status */
  status: ApplicationStatus
  /** Reason for denial if status is DENIED */
  reason_if_denied: string | null
}

File: backend/src/schemas/graph.py
"""Pydantic schemas for Neo4j graph entities.

This module defines schemas for all node types in the government processes graph,
as specified in /docs/data_model.md.

All regulatory entities (Process, Step, Requirement, Rule, DocumentType, Office, etc.)
include citation fields for source traceability.
"""

from datetime import date, datetime
from enum import Enum

from pydantic import BaseModel, Field, HttpUrl, field_validator


class ConfidenceLevel(str, Enum):
    """Confidence score for regulatory claims.

    Attributes:
        HIGH: Direct quote from official source
        MEDIUM: Inferred from official source
        LOW: Ambiguous or uncertain
    """

    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class ProcessCategory(str, Enum):
    """Categories of government processes.

    Attributes:
        PERMITS: Parking permits, building permits, etc.
        LICENSES: Professional licenses, business licenses, etc.
        BENEFITS: Government benefits programs
    """

    PERMITS = "permits"
    LICENSES = "licenses"
    BENEFITS = "benefits"


class WebResourceType(str, Enum):
    """Types of web resources.

    Attributes:
        HOW_TO: How-to guide or instructions
        PROGRAM: Program overview page
        PORTAL: Online application portal
        PDF_FORM: PDF form download
    """

    HOW_TO = "how_to"
    PROGRAM = "program"
    PORTAL = "portal"
    PDF_FORM = "pdf_form"


class ApplicationStatus(str, Enum):
    """Status of a user application.

    Attributes:
        PENDING: Application submitted, awaiting review
        APPROVED: Application approved
        DENIED: Application denied
    """

    PENDING = "pending"
    APPROVED = "approved"
    DENIED = "denied"


class ApplicationCategory(str, Enum):
    """Category of application.

    Attributes:
        NEW: New application
        RENEWAL: Renewal of existing permit/license
        REPLACEMENT: Replacement of lost/stolen permit
        RENTAL: Rental-specific application
    """

    NEW = "new"
    RENEWAL = "renewal"
    REPLACEMENT = "replacement"
    RENTAL = "rental"


class TimestampMixin(BaseModel):
    """Base mixin for timestamp fields.

    Attributes:
        created_at: When the record was created
        updated_at: When the record was last updated
    """

    created_at: datetime = Field(
        default_factory=datetime.now, description="Timestamp when record was created"
    )
    updated_at: datetime = Field(
        default_factory=datetime.now, description="Timestamp when record was last updated"
    )


class CitationMixin(BaseModel):
    """Base mixin for citation/provenance fields.

    All regulatory data must be traceable to official sources.

    Attributes:
        source_url: URL to official source document
        last_verified: Date when information was last verified
        confidence: Confidence level in this data
    """

    source_url: HttpUrl = Field(description="URL to official source (regulation, webpage, PDF)")
    last_verified: date = Field(
        description="Date when this information was last verified (YYYY-MM-DD)"
    )
    confidence: ConfidenceLevel = Field(description="Confidence level in this data")


class Process(CitationMixin, TimestampMixin):
    """High-level government service or process.

    Example: "Boston Resident Parking Permit"

    Attributes:
        process_id: Unique identifier (e.g., "boston_rpp")
        name: Display name
        description: One-sentence summary
        category: Process category (permits, licenses, benefits)
        jurisdiction: Governing authority (e.g., "City of Boston")
    """

    process_id: str = Field(min_length=1, max_length=100, description="Unique process identifier")
    name: str = Field(min_length=1, max_length=200, description="Human-readable process name")
    description: str = Field(min_length=1, description="One-sentence summary of the process")
    category: ProcessCategory = Field(description="Process category")
    jurisdiction: str = Field(
        min_length=1, max_length=200, description="Governing authority (e.g., 'City of Boston')"
    )


class Step(CitationMixin, TimestampMixin):
    """Actionable task within a process.

    Example: "Gather proof of residency"

    Attributes:
        step_id: Unique identifier (e.g., "rpp.gather_proof")
        process_id: Parent process identifier
        name: Short label for the step
        description: Full action description
        order: Sequence number for display
        estimated_time_minutes: Official time estimate
        observed_time_minutes: Median from user feedback (optional)
        cost_usd: Cost in USD if applicable
        optional: Whether step can be skipped
    """

    step_id: str = Field(min_length=1, max_length=100, description="Unique step identifier")
    process_id: str = Field(min_length=1, max_length=100, description="Parent process ID")
    name: str = Field(min_length=1, max_length=200, description="Short step label")
    description: str = Field(min_length=1, description="Full action description")
    order: int = Field(ge=1, description="Sequence number (1-indexed)")
    estimated_time_minutes: int | None = Field(
        default=None, ge=0, description="Official time estimate in minutes"
    )
    observed_time_minutes: int | None = Field(
        default=None, ge=0, description="Median time from user feedback"
    )
    cost_usd: float = Field(default=0.0, ge=0.0, description="Cost in USD")
    optional: bool = Field(default=False, description="Whether this step can be skipped")


class Requirement(CitationMixin, TimestampMixin):
    """Eligibility condition or rule.

    Example: "MA registration at Boston address"

    Attributes:
        requirement_id: Unique identifier (e.g., "rpp.req.ma_registration")
        text: Human-readable description
        fact_id: Reference to Facts Registry
        applies_to_process: Process ID this applies to
        hard_gate: Whether this blocks progress if not met
        source_section: Page/PDF section (optional)
    """

    requirement_id: str = Field(
        min_length=1, max_length=100, description="Unique requirement identifier"
    )
    text: str = Field(min_length=1, description="Human-readable requirement description")
    fact_id: str = Field(
        min_length=1,
        max_length=100,
        description="Reference to Facts Registry (e.g., 'rpp.eligibility.registration_state')",
    )
    applies_to_process: str = Field(
        min_length=1, max_length=100, description="Process ID this requirement applies to"
    )
    hard_gate: bool = Field(default=True, description="Whether this blocks progress if not met")
    source_section: str | None = Field(default=None, description="Page or PDF section reference")


class Rule(CitationMixin, TimestampMixin):
    """Atomic regulatory fact.

    Example: "Proof of residency ≤30 days"

    Attributes:
        rule_id: Unique identifier (e.g., "RPP-15-4A")
        text: Exact regulation text
        fact_id: Reference to Facts Registry
        scope: Scope of rule (general, rental, military, etc.)
        source_section: Section/page number
        effective_date: When rule took effect (optional)
    """

    rule_id: str = Field(min_length=1, max_length=100, description="Unique rule identifier")
    text: str = Field(min_length=1, description="Exact regulation text")
    fact_id: str = Field(min_length=1, max_length=100, description="Reference to Facts Registry")
    scope: str = Field(
        default="general",
        min_length=1,
        max_length=100,
        description="Scope of rule (e.g., 'general', 'rental', 'military')",
    )
    source_section: str = Field(min_length=1, description="Section/page number reference")
    effective_date: date | None = Field(default=None, description="When this rule took effect")


class DocumentType(CitationMixin, TimestampMixin):
    """Template for accepted documents.

    Example: "Utility bill ≤30 days"

    Attributes:
        doc_type_id: Unique identifier (e.g., "proof.utility_bill")
        name: Display name
        freshness_days: Maximum age in days
        name_match_required: Must match applicant name
        address_match_required: Must match Boston address
        examples: List of example issuers
    """

    doc_type_id: str = Field(
        min_length=1, max_length=100, description="Unique document type identifier"
    )
    name: str = Field(min_length=1, max_length=200, description="Human-readable document type name")
    freshness_days: int = Field(ge=0, description="Maximum age in days (e.g., 30)")
    name_match_required: bool = Field(
        default=True, description="Whether document name must match applicant name"
    )
    address_match_required: bool = Field(
        default=True, description="Whether document address must match Boston address"
    )
    examples: list[str] = Field(
        default_factory=list,
        description="List of example issuers (e.g., ['National Grid', 'Eversource'])",
    )


class Document(TimestampMixin):
    """User-provided document instance.

    Note: No citation fields - this is user data, not regulatory data.
    Documents are auto-deleted after 24 hours.

    Attributes:
        doc_id: Unique identifier (UUID)
        doc_type_id: Reference to DocumentType
        issuer: Document issuer (e.g., "National Grid")
        issue_date: Date from OCR
        name_on_doc: Name from OCR
        address_on_doc: Address from OCR
        file_ref: S3/local path (deleted after 24h)
        verified: Whether document passed validation
        validation_errors: List of validation errors if any
        deleted_at: Auto-set to created_at + 24h
    """

    doc_id: str = Field(min_length=1, description="Unique document ID (UUID)")
    doc_type_id: str = Field(min_length=1, max_length=100, description="Reference to DocumentType")
    issuer: str = Field(min_length=1, max_length=200, description="Document issuer (from OCR)")
    issue_date: date = Field(description="Issue date (from OCR)")
    name_on_doc: str = Field(
        min_length=1, max_length=200, description="Name on document (from OCR)"
    )
    address_on_doc: str = Field(min_length=1, description="Address on document (from OCR)")
    file_ref: str = Field(min_length=1, description="File path (S3/local, deleted after 24h)")
    verified: bool = Field(default=False, description="Whether document passed validation")
    validation_errors: list[str] = Field(
        default_factory=list, description="Validation errors if any"
    )
    deleted_at: datetime | None = Field(
        default=None, description="When document will be/was deleted (created_at + 24h)"
    )


class Office(CitationMixin, TimestampMixin):
    """Physical location for in-person steps.

    Example: "Office of the Parking Clerk"

    Attributes:
        office_id: Unique identifier (e.g., "parking_clerk")
        name: Office name
        address: Full street address
        room: Room number (optional)
        hours: Operating hours
        phone: Phone number (optional)
        email: Email address (optional)
    """

    office_id: str = Field(min_length=1, max_length=100, description="Unique office identifier")
    name: str = Field(min_length=1, max_length=200, description="Office name")
    address: str = Field(min_length=1, description="Full street address")
    room: str | None = Field(default=None, description="Room number")
    hours: str = Field(min_length=1, description="Operating hours (e.g., 'Mon-Fri, 9:00-4:30')")
    phone: str | None = Field(default=None, description="Phone number")
    email: str | None = Field(default=None, description="Email address")


class RPPNeighborhood(CitationMixin, TimestampMixin):
    """Boston parking neighborhood (RPP-specific).

    Attributes:
        nbrhd_id: Unique identifier (e.g., "back_bay")
        name: Display name
        auto_renew_cycle: Next audit date (optional)
        posted_streets: List of streets with RPP signage
        notes: Special rules (optional)
    """

    nbrhd_id: str = Field(
        min_length=1, max_length=100, description="Unique neighborhood identifier"
    )
    name: str = Field(min_length=1, max_length=200, description="Neighborhood display name")
    auto_renew_cycle: date | None = Field(default=None, description="Next auto-renewal audit date")
    posted_streets: list[str] = Field(
        default_factory=list, description="Street names with RPP signage"
    )
    notes: str | None = Field(default=None, description="Special rules or notes")


class WebResource(TimestampMixin):
    """Official web resource (page, PDF, portal).

    Note: No citation fields - this IS a source, not derived from one.

    Attributes:
        res_id: Unique identifier (e.g., "howto")
        title: Page title
        url: Full URL (must be unique)
        type: Resource type
        owner: Owning office/department
        last_seen: Last successful fetch date
        hash: SHA256 of content for change detection
    """

    res_id: str = Field(min_length=1, max_length=100, description="Unique resource identifier")
    title: str = Field(min_length=1, max_length=500, description="Page title")
    url: HttpUrl = Field(description="Full URL (must be unique)")
    type: WebResourceType = Field(description="Resource type")
    owner: str = Field(
        min_length=1,
        max_length=200,
        description="Owning office/department (e.g., 'Parking Clerk', 'BTD')",
    )
    last_seen: date = Field(description="Last successful fetch date")
    hash: str = Field(
        min_length=64, max_length=64, description="SHA256 hash of content for change detection"
    )

    @field_validator("hash")
    @classmethod
    def validate_hash(cls, v: str) -> str:
        """Validate that hash is a valid SHA256 hex string."""
        if not all(c in "0123456789abcdef" for c in v.lower()):
            raise ValueError("Hash must be a valid SHA256 hex string")
        return v.lower()


class Person(BaseModel):
    """User (optional, for tracking).

    Note: Minimal PII - only for Phase 2+.

    Attributes:
        person_id: Unique identifier (UUID)
        email: Hashed email address
        created_at: When user record was created
    """

    person_id: str = Field(min_length=1, description="Unique person ID (UUID)")
    email: str = Field(min_length=1, description="Hashed email address")
    created_at: datetime = Field(
        default_factory=datetime.now, description="When user record was created"
    )


class Application(TimestampMixin):
    """User's process session (Phase 2+).

    Attributes:
        app_id: Unique identifier (UUID)
        process_id: Process being applied for
        category: Application category
        submitted_on: When application was submitted (optional)
        status: Current status
        reason_if_denied: Denial reason (optional)
    """

    app_id: str = Field(min_length=1, description="Unique application ID (UUID)")
    process_id: str = Field(
        min_length=1, max_length=100, description="Process ID (e.g., 'boston_rpp')"
    )
    category: ApplicationCategory = Field(description="Application category")
    submitted_on: datetime | None = Field(
        default=None, description="When application was submitted"
    )
    status: ApplicationStatus = Field(
        default=ApplicationStatus.PENDING, description="Current application status"
    )
    reason_if_denied: str | None = Field(
        default=None, description="Reason for denial if status is DENIED"
    )


# Export all schemas for easy importing
__all__ = [
    "ConfidenceLevel",
    "ProcessCategory",
    "WebResourceType",
    "ApplicationStatus",
    "ApplicationCategory",
    "TimestampMixin",
    "CitationMixin",
    "Process",
    "Step",
    "Requirement",
    "Rule",
    "DocumentType",
    "Document",
    "Office",
    "RPPNeighborhood",
    "WebResource",
    "Person",
    "Application",
]
